File: backend/src/main/java/com/politeai/infrastructure/ai/pipeline/template/TemplateSelector.java
package com.politeai.infrastructure.ai.pipeline.template;

import com.politeai.domain.transform.model.Persona;
import com.politeai.domain.transform.model.Purpose;
import com.politeai.domain.transform.model.SituationContext;
import com.politeai.domain.transform.model.Topic;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.*;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Slf4j
@Component
@RequiredArgsConstructor
public class TemplateSelector {

    private final TemplateRegistry registry;

    public record TemplateSelectionResult(
        StructureTemplate template,
        boolean s2Enforced,
        List<StructureSection> effectiveSections
    ) {}

    // PURPOSE → template ID mapping
    private static final Map<Purpose, String> PURPOSE_TEMPLATE_MAP = Map.ofEntries(
            Map.entry(Purpose.INFO_DELIVERY, "T01_GENERAL"),
            Map.entry(Purpose.DATA_REQUEST, "T02_DATA_REQUEST"),
            Map.entry(Purpose.SCHEDULE_COORDINATION, "T04_SCHEDULE"),
            Map.entry(Purpose.APOLOGY_RECOVERY, "T05_APOLOGY"),
            Map.entry(Purpose.RESPONSIBILITY_SEPARATION, "T09_BLAME_SEPARATION"),
            Map.entry(Purpose.REJECTION_NOTICE, "T06_REJECTION"),
            Map.entry(Purpose.REFUND_REJECTION, "T11_REFUND_REJECTION"),
            Map.entry(Purpose.WARNING_PREVENTION, "T12_WARNING_PREVENTION"),
            Map.entry(Purpose.RELATIONSHIP_RECOVERY, "T10_RELATIONSHIP_RECOVERY"),
            Map.entry(Purpose.NEXT_ACTION_CONFIRM, "T01_GENERAL"),
            Map.entry(Purpose.ANNOUNCEMENT, "T07_ANNOUNCEMENT")
    );

    // CONTEXT → template ID mapping (primary context)
    private static final Map<SituationContext, String> CONTEXT_TEMPLATE_MAP = Map.ofEntries(
            Map.entry(SituationContext.REQUEST, "T02_DATA_REQUEST"),
            Map.entry(SituationContext.SCHEDULE_DELAY, "T04_SCHEDULE"),
            Map.entry(SituationContext.URGING, "T03_NAGGING_REMINDER"),
            Map.entry(SituationContext.REJECTION, "T06_REJECTION"),
            Map.entry(SituationContext.APOLOGY, "T05_APOLOGY"),
            Map.entry(SituationContext.COMPLAINT, "T09_BLAME_SEPARATION"),
            Map.entry(SituationContext.ANNOUNCEMENT, "T07_ANNOUNCEMENT"),
            Map.entry(SituationContext.FEEDBACK, "T08_FEEDBACK"),
            Map.entry(SituationContext.BILLING, "T09_BLAME_SEPARATION"),
            Map.entry(SituationContext.SUPPORT, "T05_APOLOGY"),
            Map.entry(SituationContext.CONTRACT, "T06_REJECTION"),
            Map.entry(SituationContext.RECRUITING, "T01_GENERAL"),
            Map.entry(SituationContext.CIVIL_COMPLAINT, "T09_BLAME_SEPARATION"),
            Map.entry(SituationContext.GRATITUDE, "T10_RELATIONSHIP_RECOVERY")
    );

    // Refund keywords for keyword override
    private static final Pattern REFUND_KEYWORDS = Pattern.compile(
            "환불|취소|반품|결제\\s*취소|카드\\s*취소|refund|cancel"
    );

    public TemplateSelectionResult selectTemplate(
            Persona persona,
            List<SituationContext> contexts,
            Topic topic,
            Purpose purpose,
            LabelStats labelStats,
            String maskedText) {

        String templateId;

        // 1. PURPOSE provided → direct mapping
        if (purpose != null) {
            templateId = PURPOSE_TEMPLATE_MAP.getOrDefault(purpose, "T01_GENERAL");
            log.info("[TemplateSelector] Selected by PURPOSE: {} → {}", purpose, templateId);
        }
        // 2. Primary CONTEXT → mapping
        else if (!contexts.isEmpty()) {
            SituationContext primaryContext = contexts.get(0);
            templateId = CONTEXT_TEMPLATE_MAP.getOrDefault(primaryContext, "T01_GENERAL");
            log.info("[TemplateSelector] Selected by CONTEXT: {} → {}", primaryContext, templateId);
        }
        // 3. Default
        else {
            templateId = "T01_GENERAL";
            log.info("[TemplateSelector] Default template: T01_GENERAL");
        }

        // 4. Topic override: REFUND_CANCEL + rejection-like → T11
        if (topic == Topic.REFUND_CANCEL && isRejectionLike(purpose, contexts)) {
            templateId = "T11_REFUND_REJECTION";
            log.info("[TemplateSelector] Topic override → T11_REFUND_REJECTION");
        }

        // 5. Keyword override: refund keywords + rejection labels
        if (!templateId.equals("T11_REFUND_REJECTION") && maskedText != null
                && REFUND_KEYWORDS.matcher(maskedText).find()
                && (labelStats.hasNegativeFeedback() || isRejectionLike(purpose, contexts))) {
            templateId = "T11_REFUND_REJECTION";
            log.info("[TemplateSelector] Keyword override → T11_REFUND_REJECTION");
        }

        StructureTemplate template = registry.get(templateId);

        // 6. S2 enforcement: ACCOUNTABILITY or NEGATIVE_FEEDBACK → inject S2 if missing
        boolean s2Enforced = false;
        List<StructureSection> sections = new ArrayList<>(template.sectionOrder());
        if ((labelStats.hasAccountability() || labelStats.hasNegativeFeedback())
                && !sections.contains(StructureSection.S2_OUR_EFFORT)) {
            // Insert S2 after S1 (or after S0 if S1 not present)
            int insertIdx = sections.indexOf(StructureSection.S1_ACKNOWLEDGE);
            if (insertIdx < 0) insertIdx = sections.indexOf(StructureSection.S0_GREETING);
            sections.add(insertIdx + 1, StructureSection.S2_OUR_EFFORT);
            s2Enforced = true;
            log.info("[TemplateSelector] S2 enforced for ACCOUNTABILITY/NEGATIVE_FEEDBACK");
        }

        // 7. Apply persona skip rules → produce effectiveSections
        List<StructureSection> effective = applySkipRules(sections, template, persona);

        return new TemplateSelectionResult(template, s2Enforced, effective);
    }

    private boolean isRejectionLike(Purpose purpose, List<SituationContext> contexts) {
        if (purpose == Purpose.REJECTION_NOTICE || purpose == Purpose.REFUND_REJECTION) return true;
        return contexts.contains(SituationContext.REJECTION);
    }

    private List<StructureSection> applySkipRules(List<StructureSection> sections,
                                                    StructureTemplate template,
                                                    Persona persona) {
        if (template.personaSkipRules() == null || !template.personaSkipRules().containsKey(persona)) {
            return sections;
        }

        StructureTemplate.SectionSkipRule rule = template.personaSkipRules().get(persona);
        return sections.stream()
                .filter(s -> !rule.skipSections().contains(s))
                .collect(Collectors.toList());
    }
}

File: app/pipeline/template/template_selector.py
"""Template selector — PURPOSE→CONTEXT→keyword→S2 enforcement→skip rules."""

import logging
import re
from dataclasses import dataclass

from app.models.domain import LabelStats
from app.models.enums import Purpose, Topic
from app.pipeline.template.structure_template import StructureSection, StructureTemplate
from app.pipeline.template.template_registry import TemplateRegistry

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class TemplateSelectionResult:
    template: StructureTemplate
    s2_enforced: bool
    effective_sections: list[StructureSection]


# PURPOSE → template ID mapping
_PURPOSE_TEMPLATE_MAP: dict[Purpose, str] = {
    Purpose.INFO_DELIVERY: "T01_GENERAL",
    Purpose.DATA_REQUEST: "T02_DATA_REQUEST",
    Purpose.SCHEDULE_COORDINATION: "T04_SCHEDULE",
    Purpose.APOLOGY_RECOVERY: "T05_APOLOGY",
    Purpose.RESPONSIBILITY_SEPARATION: "T09_BLAME_SEPARATION",
    Purpose.REJECTION_NOTICE: "T06_REJECTION",
    Purpose.REFUND_REJECTION: "T11_REFUND_REJECTION",
    Purpose.WARNING_PREVENTION: "T12_WARNING_PREVENTION",
    Purpose.RELATIONSHIP_RECOVERY: "T10_RELATIONSHIP_RECOVERY",
    Purpose.NEXT_ACTION_CONFIRM: "T01_GENERAL",
    Purpose.ANNOUNCEMENT: "T07_ANNOUNCEMENT",
}

_REFUND_KEYWORDS = re.compile(r"환불|취소|반품|결제\s*취소|카드\s*취소|refund|cancel")


def select_template(
    registry: TemplateRegistry,
    topic: Topic | None,
    purpose: Purpose | None,
    label_stats: LabelStats,
    masked_text: str | None,
) -> TemplateSelectionResult:
    # 1. PURPOSE provided → direct mapping
    if purpose is not None:
        template_id = _PURPOSE_TEMPLATE_MAP.get(purpose, "T01_GENERAL")
        logger.info("[TemplateSelector] Selected by PURPOSE: %s → %s", purpose, template_id)
    # 2. Default
    else:
        template_id = "T01_GENERAL"
        logger.info("[TemplateSelector] Default template: T01_GENERAL")

    # 3. Topic override: REFUND_CANCEL + rejection-like → T11
    if topic == Topic.REFUND_CANCEL and _is_rejection_like(purpose):
        template_id = "T11_REFUND_REJECTION"
        logger.info("[TemplateSelector] Topic override → T11_REFUND_REJECTION")

    # 4. Keyword override: refund keywords + rejection labels
    if (
        template_id != "T11_REFUND_REJECTION"
        and masked_text is not None
        and _REFUND_KEYWORDS.search(masked_text)
        and (label_stats.has_negative_feedback or _is_rejection_like(purpose))
    ):
        template_id = "T11_REFUND_REJECTION"
        logger.info("[TemplateSelector] Keyword override → T11_REFUND_REJECTION")

    template = registry.get(template_id)

    # 5. S2 enforcement: ACCOUNTABILITY or NEGATIVE_FEEDBACK → inject S2 if missing
    s2_enforced = False
    sections = list(template.section_order)
    if (
        (label_stats.has_accountability or label_stats.has_negative_feedback)
        and StructureSection.S2_OUR_EFFORT not in sections
    ):
        insert_idx = -1
        if StructureSection.S1_ACKNOWLEDGE in sections:
            insert_idx = sections.index(StructureSection.S1_ACKNOWLEDGE)
        elif StructureSection.S0_GREETING in sections:
            insert_idx = sections.index(StructureSection.S0_GREETING)
        sections.insert(insert_idx + 1, StructureSection.S2_OUR_EFFORT)
        s2_enforced = True
        logger.info("[TemplateSelector] S2 enforced for ACCOUNTABILITY/NEGATIVE_FEEDBACK")

    return TemplateSelectionResult(
        template=template,
        s2_enforced=s2_enforced,
        effective_sections=sections,
    )


def _is_rejection_like(purpose: Purpose | None) -> bool:
    return purpose in (Purpose.REJECTION_NOTICE, Purpose.REFUND_REJECTION)
